Tests: Various race fixes and move TestFixtureToDataHandler

gbjk opened the pull request:
* Fix a instance pair manager race intermittently causing Bitfinex and Kraken to fail
* Fix PairsManager.Load breaking matcher
* Fix UpdatePairsOnce not working across instances
* Move and Simplify FixtureToDataHandler
* Rename PairStore.copy to PairStore.clone and replace various builtin `copy` with slices.Clone
* Add Fullstore.clone and PairFormat.clone

## Type of change

- [x] Test fixes and improvements

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run

shazbert commented on currency/manager.go:
```diff
@@ -208,8 +205,7 @@ func (p *PairsManager) StorePairs(a asset.Item, pairs Pairs, enabled bool) error
 	// remove the entire stored list if needed.
-	cpy := make(Pairs, len(pairs))
-	copy(cpy, pairs)
+	cpy := slices.Clone(pairs)
```
suggestion: can rm the var and comment and inline to assignment.

gbjk replied:
Yep. In doing so I'm going to remove the comment, which needed rewording at best, but feels implicit to me.

Fixed ef4e43c

shazbert commented on currency/pair.go:
```diff
@@ -118,6 +118,16 @@ func (f PairFormat) Format(pair Pair) string {
+	if f == nil {
+		return nil
+	}
+	c := &PairFormat{}
+	*c = *f
+	return c
```
```suggestion
	c := *f
	return &c
```

gbjk replied:
🤦 
Fixed d6480c1

shazbert commented on currency/manager.go:
```diff
@@ -435,34 +431,27 @@ func (p *PairsManager) SetAssetEnabled(a asset.Item, enabled bool) error {
+	p.matcher = map[key]*Pair{}
+
+	// Re-index matcher
+	for a, fs := range p.Pairs {
```
Suggestion: This is done a few times could make a method for it.

gbjk replied:
This exposed that the matcher index was never flushed when storing pairs.
We have three options:
1. Blow away the whole index each time regardless of asset updated
2. Restructure the key to make it fast to blow away just one asset.
3. Walk the current map and delete by asset first

Since store pair is relatively infrequent, I've opted to blow away the whole index each time, for simplicity.

Fixed 9005629

shazbert replied:
🔥 Nice 👍

shazbert commented on currency/manager.go:
```diff
@@ -539,21 +528,18 @@ func (fs FullStore) MarshalJSON() ([]byte, error) {
-
-	avail := make(Pairs, len(ps.Available))
-	copy(avail, ps.Available)
+	enabled := slices.Clone(ps.Enabled)
+	avail := slices.Clone(ps.Available)
```
Suggestion: can remove these declarations and use them in the Pairs Store allocation and below you can remove rFmt and cFmt and just use the methods you made.

gbjk replied:
Fixed 30361ff1

shazbert reviewed (changes requested):
Looks good to me, only some pedantic things, that you can resolve at your choosing. Nice.

gbjk commented on currency/manager.go:
```diff
@@ -157,8 +155,7 @@ func (p *PairsManager) GetPairs(a asset.Item, enabled bool) (Pairs, error) {
-	copy(enabledPairs, pairStore.Enabled)
+	enabledPairs := slices.Clone(pairStore.Enabled)
 
 	err := pairStore.Available.ContainsAll(pairStore.Enabled, true)
```
Fixed 23773fd

shazbert reviewed (changes requested):
Just some 🏃 racey features left over [here](https://github.com/thrasher-corp/gocryptotrader/actions/runs/8962919898/job/24612474164#step:9:511)

gbjk commented:
Fixed with 2e9e7d67 and 1481244a
In both cases I've shied away from going and changing everything to remove the (very) thin wrappers around CurrencyManager methods
Also haven't gone hunting for other examples of intrusion (yet)

shazbert commented on currency/manager.go:
```diff
@@ -186,6 +177,43 @@ func (p *PairsManager) StoreFormat(a asset.Item, pFmt *PairFormat, config bool)
 }
 
+// GetFormat returns the pair format in a concurrent safe manner
+func (p *PairsManager) GetFormat(a asset.Item, r bool) (PairFormat, error) {
```
Can you expand the `r` name to something else a little easier for somebody to understand? This also needs a test.
 
suggestion: Also can consolidate this down a little bit, example below, it calls p.Get just to verify asset in every case, would need to change the error to something more generic for this to work:

```go
	ps, err := p.Get(a)
	if err != nil {
		return EMPTYFORMAT, err
	}
	var outbound *PairFormat
	if p.UseGlobalFormat {
		outbound = p.ConfigFormat
		if r {
			outbound = p.RequestFormat
		}
	} else {
		outbound = ps.ConfigFormat
		if r {
			outbound = ps.RequestFormat
		}
	}
	if outbound == nil {
		return EMPTYFORMAT, fmt.Errorf("global: %v request %v: %w", p.UseGlobalFormat, r, ErrAssetConfigFormatIsNil)
	}
	return *outbound, nil

```

gbjk replied:
💯 can do
I had the same thoughts myself, started to do exactly this, but then thought I should rein myself in.
You've given me carte blanche so I'll have at it.
I wanted to simplify the error, as well. Seemed like overkill.

gbjk replied:
Fixed 907d016

Thanks for the suggested simplification.
I ensured that we don't try to get the asset if we're using Global, and I kept a much simpler (existing) error for nil, which should be an edge case anyway.

shazbert commented on exchanges/gateio/gateio.go:
```diff
@@ -3753,8 +3753,12 @@ func (g *Gateio) IsValidPairString(currencyPair string) bool {
-		result := strings.Split(currencyPair, g.CurrencyPairs.RequestFormat.Delimiter)
+	pf, err := g.CurrencyPairs.GetFormat(asset.Spot, true)
+	if err != nil {
+		return false
```
I am just pointing out that something ungodly is going down if this is erroring, so just a _suggestion_ to return an error along with the bool. But up to you.

gbjk replied:
Thought about it.
But I think that shouldn't be the responsibility of this function, and I want to keep `IsXYZ` functions slim.
According to the data we have, this is definitely *not* a valid string 😭 😂

shazbert reviewed (changes requested):
Good stuff, thanks for addressing my change requests. There's a couple of error on the [CI build](https://github.com/thrasher-corp/gocryptotrader/actions/runs/8965928427/job/24620397195?pr=1534#step:9:341)

shazbert commented on currency/manager_test.go:
```diff
@@ -857,3 +857,60 @@ func TestPairManagerSetDelimitersFromConfig(t *testing.T) {
+		UseGlobalFormat: true,
+		ConfigFormat: &PairFormat{
+			Uppercase: true,
+			Delimiter: "🦄",
```
🔥

shazbert reviewed (changes requested):
Changes look good 👍 can proceed with rebase.

gbjk commented:
@shazbert Following discussion with @thrasher- I've switched it back to using a new `PairsManager.IsAssetSupported`.

Fixed 93fee81

Everything else was a rebase of what you've already seen.

shazbert reviewed (approved):
Thanks for the changes. tACK! 🚀

shazbert reviewed (approved):
Changes ACK.

gloriousCode commented on currency/manager.go:
```diff
@@ -539,39 +538,30 @@ func (fs FullStore) MarshalJSON() ([]byte, error) {
+func (ps *PairStore) clone() *PairStore {
 	if ps == nil {
-		return nil, errPairStoreIsNil
+		return nil
```
The removal of the error here and making returning nil an acceptable situation is something that could lead to panics as not much code is prepared for that. 

At present it looks like only a botched test via `Load` will lead to this which is acceptable and addressable by the offending user, but I still need to point out because 🦕

gbjk replied:
Understood.

gloriousCode reviewed (approved):
tACK!

thrasher- reviewed (approved):
tACK, nice work! NOTE: BTSE issues are being worked on in a separate branch.
